### ISSUE TITLE
Seed only when I am the only seeder

### ISSUE THREAD

**Issue by mitar** [cenkalti/rain#182]
I could not find a way to fine-tune seeding strategy? I would like to be configure seeding only when I am the only seeder, so to seed rare files, but not seed otherwise. Like an archive for less popular torrents. In library mode.

I would guess I would need a notify channel to be informed about all changes to peers and when there is no peer which seeds, I start seeding? But I am not sure if it is really possible to get full seeding information for a torrent as a client? Probably only trackers have full information?

**Comment by cenkalti** [comment]
There isn't built-in support for custom storage implementations. However, I've made a small change to make it easier to modify in the code and build yourself.

Storage appears as an interface to the client:
https://github.com/cenkalti/rain/blob/master/internal/storage/storage.go

If you are willing to write some code, this line is where you should start:
https://github.com/cenkalti/rain/blob/e8a4831e39945901c8c2e456525992407dcdeb9e/torrent/session.go#L448-L450

Please let me know if you have questions.

**Comment by cenkalti** [comment]
Because of the current design, you can't see the number of seeders without starting the torrent. To find out the seeder count, the torrent client needs to send an announce request first. I'm not sure how to get the seeder count for stopped torrents.

**Comment by mitar** [comment]
> Because of the current design, you can't see the number of seeders without starting the torrent. To find out the seeder count, the torrent client needs to send an announce request first. I'm not sure how to get the seeder count for stopped torrents.

But can one the torrent client announce the torrent, but just do not upload any data until it is the only seed?

**Comment by cenkalti** [comment]
I don't think this is the right way to implement it. Imagine two clients act like this and won't upload until they're the only seeder. A third peer connecting as a leecher wouldn't download the torrent because the two seeders won't upload, each thinking the other will.

There is a method to get seeder statistics from a tracker. It's called scraping, but it hasn't been implemented in Rain yet. I think it can be added for stopped torrents too.

I've gathered some resources for future implementation:
- https://en.wikipedia.org/wiki/Tracker_scrape
- https://web.archive.org/web/20220216014242/https://wiki.theory.org/index.php/BitTorrentSpecification#Tracker_.27scrape.27_Convention
- https://web.archive.org/web/20230315105013/https://wiki.vuze.com/w/Scrape
- https://www.bittorrent.org/beps/bep_0048.html (HTTP scrape)
- https://www.bittorrent.org/beps/bep_0015.html (UDP scrape)
- https://www.bittorrent.org/beps/bep_0033.html (DHT scrape)
- https://github.com/qbittorrent/qBittorrent/issues/18614 (similar request in another client)

If we just implement `func Torrent.Scrape() ScrapeResult` you can write your own loop to have a custom logic on when to start/stop torrents.

**Comment by mitar** [comment]
Great point! Thanks.

**Comment by mitar** [comment]
Unstale.

**Comment by mitar** [comment]
Unstale.
